### PR TITLE
MTDSA-15014: Added the remaining fields to feature switch functionality

### PR DESCRIPTION
**Pull request by abigail-houston-hmrc**

### CODE REVIEW

File: app/v3/models/response/retrieveCalculation/RetrieveCalculationResponse.scala [hunk 1]
```diff
@@ -45,6 +47,49 @@ case class RetrieveCalculationResponse(
     copy(calculation = updatedCalculation)
   }
 
+  def withoutOffPayrollWorker: RetrieveCalculationResponse = {
```
Comment (tom-lloyd-hmrc) [inline app/v3/models/response/retrieveCalculation/RetrieveCalculationResponse.scala:50]: It's a little hard to follow the logic because of how many nested layers are involved, I wonder if it'd be neater to put the logic for the nested models inside methods in those models too, so you'd end up with a bunch of functions that look like:

```
  def withoutOffPayrollWorker: RetrieveCalculationResponse = copy(calculation = calculation.map(_.withoutOffPayrollWorker).filter(_.isDefined))
```

File: app/v3/models/response/retrieveCalculation/calculation/employmentAndPensionsIncome/EmploymentAndPensionsIncome.scala [hunk 2]
```diff
@@ -22,7 +22,28 @@ case class EmploymentAndPensionsIncome(totalPayeEmploymentAndLumpSumIncome: Opti
...
+                                       employmentAndPensionsIncomeDetail: Option[Seq[EmploymentAndPensionsIncomeDetail]]) {
+
+  val isDefined: Boolean =
+    !(totalPayeEmploymentAndLumpSumIncome.isEmpty &&
+      totalOccupationalPensionIncome.isEmpty &&
+      totalBenefitsInKind.isEmpty &&
+      tipsIncome.isEmpty &&
+      employmentAndPensionsIncomeDetail.isEmpty)
+
+  def withoutOffPayrollWorker: EmploymentAndPensionsIncome = {
```
Comment (tom-lloyd-hmrc) [inline app/v3/models/response/retrieveCalculation/calculation/employmentAndPensionsIncome/EmploymentAndPensionsIncome.scala:34]: You could add further add `withoutOffPayrollWorker` to `EmploymentAndPensionsIncomeDetail` to simplify this to:

```
  def withoutOffPayrollWorker: EmploymentAndPensionsIncome = copy(
    employmentAndPensionsIncomeDetail = employmentAndPensionsIncomeDetail.map(_.map(_.withoutOffPayrollWorker)).filter(_.nonEmpty)
  )
```

File: app/v3/models/response/retrieveCalculation/calculation/taxCalculation/Nics.scala [hunk 3]
```diff
@@ -24,7 +24,20 @@ case class Nics(
     nic2NetOfDeductions: Option[BigDecimal],
     nic4NetOfDeductions: Option[BigDecimal],
     totalNic: Option[BigDecimal]
-)
+) {
+
+  def isDefined = !(
```
Comment (tom-lloyd-hmrc) [inline app/v3/models/response/retrieveCalculation/calculation/taxCalculation/Nics.scala:29]: Intellij complains about the lack of type annotation (`: Boolean`) here.

File: test/v3/models/response/retrieveCalculation/CalculationFixture.scala [hunk 4]
```diff
@@ -44,7 +47,57 @@ trait CalculationFixture {
     topSlicingRelief = None,
     reliefsClaimed = None
   )
+
+  val class2Nics: Class2Nics                                 = Class2Nics(None, None, None, None, None, true, Some(true), None)
```
Comment (tom-lloyd-hmrc) [inline test/v3/models/response/retrieveCalculation/CalculationFixture.scala:51]: It would probably be better to use named arguments for all fixtures, as it is not clear what fields are being set here.

File: app/v3/controllers/RetrieveCalculationController.scala [hunk 5]
```diff
@@ -85,4 +81,12 @@ class RetrieveCalculationController @Inject() (val authService: EnrolmentsAuthSe
 
     }
 
+  private def updateModel(response: RetrieveCalculationResponse): RetrieveCalculationResponse =
+    featureSwitches.isTaxYearSpecificApiEnabled match {
```
Comment (mattstephens-hmrc) [inline app/v3/controllers/RetrieveCalculationController.scala:85]: I think this would be a bit clearer as:
```
    def without = response.withoutBasicExtension.withoutTotalAllowanceAndDeductions

    def without = response.withoutBasicExtension.withoutTotalAllowanceAndDeductions

    isTaxYearSpecificApiEnabled match {
      case true if (!isR8bSpecificApiEnabled) => without.withoutOffPayrollWorker
      case false                              => without.withoutUnderLowerProfitThreshold.withoutOffPayrollWorker
      case true                               => response
    }
```
and put `import featureSwitches.{isR8bSpecificApiEnabled, isTaxYearSpecificApiEnabled}` at the top, under the `private val featureSwitches`

File: test/v3/models/response/retrieveCalculation/CalculationSpec.scala [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package v3.models.response.retrieveCalculation
+
+import support.UnitSpec
+import v3.models.response.retrieveCalculation.calculation.Calculation
+
+class CalculationSpec extends UnitSpec with CalculationFixture {
+
+  "update model" when {
```
Comment (mattstephens-hmrc) [inline test/v3/models/response/retrieveCalculation/CalculationSpec.scala:24]: there should be a "should" in there somewhere, maybe just `"update model" should {`

File: test/v3/models/response/retrieveCalculation/CalculationSpec.scala [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    "removeTotalAllowanceAndDeductions is called" in {
+      val calculation: Calculation = calcWithoutEndOfYearEstimate
+      minimalCalculationResponse.withoutTotalAllowanceAndDeductions shouldBe minimalCalculationResponse.copy(calculation = Some(calculation))
+    }
+    "removeBasicExtension is called" in {
```
Comment (mattstephens-hmrc) [inline test/v3/models/response/retrieveCalculation/CalculationSpec.scala:29]: the method names are "withoutXYZ", so these ought to match

File: test/v3/models/response/retrieveCalculation/CalculationSpec.scala [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      val calculation: Calculation = calcWithoutBasicExtension
```
Comment (mattstephens-hmrc) [inline test/v3/models/response/retrieveCalculation/CalculationSpec.scala:30]: If I remove the `.filter(_.isDefined)` bit in the method, the tests still all pass; I'm wondering if the filter thing does anything? If so, there ought to be a test case to cover it

Comment (mattstephens-hmrc) [reply test/v3/models/response/retrieveCalculation/CalculationSpec.scala:30]: (same for the other two)

File: test/v3/models/response/retrieveCalculation/CalculationSpec.scala [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v3.models.response.retrieveCalculation
+
+import support.UnitSpec
+import v3.models.response.retrieveCalculation.calculation.Calculation
+
+class CalculationSpec extends UnitSpec with CalculationFixture {
```
Comment (mattstephens-hmrc) [inline test/v3/models/response/retrieveCalculation/CalculationSpec.scala:22]: It does make sense for this to be separate from RetrieveCalculationResponseSpec, but it isn't really testing Calculation as the methods it's testing are in RetrieveCalculationResponse.  How about `RetrieveCalculationResponseWithoutSpec`?

File: app/v3/models/response/retrieveCalculation/calculation/employmentAndPensionsIncome/EmploymentAndPensionsIncomeDetail.scala [hunk 10]
```diff
@@ -33,7 +33,28 @@ case class EmploymentAndPensionsIncomeDetail(incomeSourceId: Option[String],
                                              disguisedRemuneration: Option[Boolean],
                                              lumpSums: Option[LumpSums],
                                              studentLoans: Option[StudentLoans],
-                                             benefitsInKind: Option[BenefitsInKind])
+                                             benefitsInKind: Option[BenefitsInKind]) {
+
+  val isDefined: Boolean =
```
Comment (mattstephens-hmrc) [inline app/v3/models/response/retrieveCalculation/calculation/employmentAndPensionsIncome/EmploymentAndPensionsIncomeDetail.scala:38]: feels like there ought to be a unit test for isDefined?  Would be a new CalculationSpec, can use the emptyCalculation fixture + some other, for the true & false results

File: app/v3/models/response/retrieveCalculation/RetrieveCalculationResponse.scala [hunk 11]
```diff
@@ -33,29 +33,14 @@ case class RetrieveCalculationResponse(
...
-      .map { calc =>
-        val updatedReliefs = calc.reliefs
-          .map(_.copy(basicRateExtension = None))
-          .filter(_.isDefined)
-        calc.copy(reliefs = updatedReliefs)
-      }
-      .filter(_.isDefined)
-    copy(calculation = updatedCalculation)
-  }
+  def withoutBasicExtension: RetrieveCalculationResponse   = copy(calculation = calculation.map(_.withoutBasicExtension).filter(_.isDefined))
```
Comment (mattstephens-hmrc) [inline app/v3/models/response/retrieveCalculation/RetrieveCalculationResponse.scala:36]: tbh I think the code could be simplified/made a bit clearer, if there's an apply() function that checks calculation.isDefined when the RetrieveCalculationResponse is created, so the response only ever has Some(non-empty Calculation) or None

Comment (mattstephens-hmrc) [reply app/v3/models/response/retrieveCalculation/RetrieveCalculationResponse.scala:36]: Basically this, in `object RetrieveCalculationResponse`:
```
  def apply(metadata: Metadata, inputs: Inputs, maybeCalc: Option[Calculation], maybeMessages: Option[Messages]): RetrieveCalculationResponse = {
    new RetrieveCalculationResponse(
      metadata,
      inputs,
      calculation = if (maybeCalc.exists(_.isDefined)) maybeCalc else None,
      maybeMessages
    )
  }
```

Comment (mattstephens-hmrc) [reply app/v3/models/response/retrieveCalculation/RetrieveCalculationResponse.scala:36]: Then the "copy" methods can be like:
```
def withoutBasicExtension: RetrieveCalculationResponse = RetrieveCalculationResponse(metadata, inputs, calculation, messages)
```

File: test/v3/models/response/retrieveCalculation/CalculationSpec.scala (hunk 8 above)
Comment (mattstephens-hmrc) [reply test/v3/models/response/retrieveCalculation/CalculationSpec.scala:30]: I'd suggest structuring the tests like this (given the above new "apply" function):
```
  "withoutTotalAllowanceAndDeductions" when {
    "calculation.isDefined returns true" should {
      "return the response with the calculation" in {
        val calculation = calcWithoutEndOfYearEstimate
        val result      = minimalCalculationResponse.withoutTotalAllowanceAndDeductions
        result shouldBe minimalCalculationResponse.copy(calculation = Some(calculation))
      }
    }

    "calculation.isDefined returns false" should {
      "return the response with no calculation" in {
        val response = minimalCalculationResponse.copy(calculation = Some(emptyCalculation))
        val result   = response.withoutTotalAllowanceAndDeductions
        result shouldBe minimalCalculationResponse.copy(calculation = None)
      }
    }
  }
```

File: test/v3/models/response/retrieveCalculation/RetrieveCalculationResponseWithoutSpec.scala [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+ * limitations under the License.
+ */
+
+package v3.models.response.retrieveCalculation
+
+import support.UnitSpec
+
+class RetrieveCalculationResponseWithoutSpec extends UnitSpec with CalculationFixture {
+
+  "update model" should {
```
Comment (mattstephens-hmrc) [inline test/v3/models/response/retrieveCalculation/RetrieveCalculationResponseWithoutSpec.scala:23]: this doesn't need the outer "should" (too many shoulds!), can just be deleted

**Review by mattstephens-hmrc — APPROVED** [review]
